src/client.ts

```diff
@@ -19,103 +25,179 @@ export class APIClient {
-    // where each resource is processed and dropped from memory.
-
-    const users: AcmeUser[] = [
-      {
-        id: 'acme-user-1',
-        name: 'User One',
-      },
-      {
-        id: 'acme-user-2',
-        name: 'User Two',
-      },
-    ];
-
-    for (const user of users) {
+    const endpoint = this.buildEndpoint(corpName, '/users');
+
+    const { data } = await this.getJsonWithRetries(endpoint, {
+      headers: this.generateHeaders(),
+    });
+
+    for (const user of data) {
       await iteratee(user);
     }
   }
 
+  private buildEndpoint(corp: string, path?: string) {
+    if (!corp) {
+      throw new IntegrationError({
+        message: `Parameter 'corp' is required, but received ${corp}`,
+        code: 'INVALID_PARAM_ERROR',
+      });
+    }
+
+    if (path) {
+      return `${BASE_URI}/corps/${corp}${path}`;
+    } else {
+      return `${BASE_URI}/corps/${corp}`;
+    }
+  }
+
+  private generateHeaders() {
+    const { apiUser, apiToken } = this.config;
+
+    return {
+      'x-api-user': apiUser,
+      'x-api-token': apiToken,
+      'Content-Type': 'application/json',
+    };
+  }
+
+  private async getJsonWithRetries(endpoint, options, retries = 0) {
+    try {
+      return await this.getJson(endpoint, options);
+    } catch (error) {
+      if ([429, 503, 504].includes(error.status) && retries < 5) {
+        // TODO: add delay before retry (sp - 1/2022)
+        return await this.getJsonWithRetries(endpoint, options, ++retries);
+      } else {
+        console.error(error);
+
+        throw new IntegrationProviderAPIError({
+          cause: error,
+          endpoint,
+          status: error.status,
+          statusText: error.statusText,
+        });
+      }
+    }
+  }
+
   /**
-   * Iterates each group resource in the provider.
-   *
-   * @param iteratee receives each resource to produce entities/relationships
+   * Uses nodejs https.get method to make HTTP requests.
+   * Parses JSON result, returning the data.
+   * @param endpoint
+   * @param options
+   * @returns Promise
    */
-  public async iterateGroups(
-    iteratee: ResourceIteratee<AcmeGroup>,
-  ): Promise<void> {
-    // TODO paginate an endpoint, invoke the iteratee with each record in the
-    // page
-    //
-    // The provider API will hopefully support pagination. Functions like this
-    // should maintain pagination state, and for each page, for each record in
-    // the page, invoke the `ResourceIteratee`. This will encourage a pattern
-    // where each resource is processed and dropped from memory.
-
-    const groups: AcmeGroup[] = [
-      {
-        id: 'acme-group-1',
-        name: 'Group One',
-        users: [
-          {
-            id: 'acme-user-1',
-          },
-        ],
-      },
-    ];
-
-    for (const group of groups) {
-      await iteratee(group);
-    }
+  private getJson(
+    endpoint: string,
+    options: https.RequestOptions,
+  ): Promise<any> {
```

Comment:
if the JSON coming back from signalsciences has some common response structure, we should make a type for it and include in an explicit return type:
```
{
status: number,
data: `someTypeYouDefine`
}
```